File: apps/desktop/src/lib/timeline/segments.ts
/**
 * Timeline segment model: the basis for split + ripple-delete editing.
 *
 * Kept content = `[trimStart, trimEnd]` minus the union of `cuts`. `splitPoints`
 * (original-recording seconds) subdivide that kept content into addressable
 * **segments** without removing anything; a split only matters once one of the
 * segments is deleted (which becomes a `manual` cut). All functions are pure.
 * See ./cuts.ts for the time-remapping that closes gaps left by deletes.
 */

import { normalizeCuts, type TimelineCut } from "./cuts";

/** Tolerance for treating two times as the same boundary. Matches cuts.ts. */
const EPS = 1e-4;

/** A contiguous kept sub-range on the original-recording timeline. */
export interface Segment {
	/** Start on the original-recording timeline (seconds). */
	start: number;
	/** End on the original-recording timeline (seconds). */
	end: number;
	/** Position in the ordered, left-to-right segment list. */
	index: number;
}

export interface ClipShape {
	/** Kept-region start (seconds). */
	trimStart: number;
	/** Kept-region end (seconds). */
	trimEnd: number;
	cuts: TimelineCut[];
	/** Division markers, original seconds. Order/duplicates don't matter. */
	splitPoints: number[];
}

/**
 * Derive the ordered list of kept segments. The kept region `[trimStart,
 * trimEnd]` has the cuts removed, then each surviving interval is sliced at any
 * split points that fall strictly inside it. Zero-length results are dropped,
 * and stray split points (outside the clip or inside a cut) are ignored, so
 * the result stays valid even after a trim or cut moves under a split.
 */
export function deriveSegments(shape: ClipShape): Segment[] {
	const { trimStart, trimEnd, splitPoints } = shape;
	if (trimEnd - trimStart <= EPS) return [];

	// Kept intervals = [trimStart, trimEnd] minus the normalized cuts.
	const cuts = normalizeCuts(shape.cuts).filter(
		(c) => c.end > trimStart && c.start < trimEnd,
	);
	const kept: Array<{ start: number; end: number }> = [];
	let cursor = trimStart;
	for (const c of cuts) {
		const cutStart = Math.max(c.start, trimStart);
		const cutEnd = Math.min(c.end, trimEnd);
		if (cutStart - cursor > EPS) kept.push({ start: cursor, end: cutStart });
		cursor = Math.max(cursor, cutEnd);
	}
	if (trimEnd - cursor > EPS) kept.push({ start: cursor, end: trimEnd });

	// Slice each kept interval at the split points strictly inside it.
	const segments: Segment[] = [];
	for (const interval of kept) {
		const inside = splitPoints
			.filter((p) => p > interval.start + EPS && p < interval.end - EPS)
			.sort((a, b) => a - b);
		let from = interval.start;
		for (const p of inside) {
			segments.push({ start: from, end: p, index: segments.length });
			from = p;
		}
		segments.push({ start: from, end: interval.end, index: segments.length });
	}
	return segments;
}

/**
 * A collapsed cut between two kept segments. On the output (post-cut) timeline
 * the removed span has zero width, so the editor renders it as a single seam
 * marker at `gapStart` (== the previous segment's end). `removed` is how much
 * original time was taken out, shown in the restore tooltip.
 */
export interface Seam {
	/** Original time where the gap begins (previous segment's end). */
	gapStart: number;
	/** Original time where the gap ends (next segment's start). */
	gapEnd: number;
	/** Seconds of original content removed across the gap. */
	removed: number;
}

/**
 * Derive the seams between adjacent kept segments: every place a cut was
 * ripple-removed, collapsing two segments together. Segments that merely *touch*
 * (a split, no removed time) yield no seam.
 */
export function deriveSeams(segments: Segment[]): Seam[] {
	const seams: Seam[] = [];
	for (let i = 0; i < segments.length - 1; i++) {
		const gap = segments[i + 1].start - segments[i].end;
		if (gap > EPS) {
			seams.push({
				gapStart: segments[i].end,
				gapEnd: segments[i + 1].start,
				removed: gap,
			});
		}
	}
	return seams;
}

/**
 * The segment containing original time `t`, or null if `t` is in a trimmed-off
 * region or inside a cut. On an exact internal boundary the playhead belongs to
 * the segment to its right (NLE convention); at the very end it belongs to the
 * last segment.
 */
export function segmentAt(segments: Segment[], t: number): Segment | null {
	for (const seg of segments) {
		if (t >= seg.start - EPS && t < seg.end - EPS) return seg;
	}
	const last = segments[segments.length - 1];
	if (last && Math.abs(t - last.end) <= EPS) return last;
	return null;
}

/**
 * Add a split at original time `t`. Returns the new sorted split-point array,
 * or null when the split is a no-op or illegal: at/outside the clip edges,
 * inside a cut, or coincident with an existing split or boundary.
 */
export function planSplit(t: number, shape: ClipShape): number[] | null {
	const { trimStart, trimEnd, splitPoints } = shape;
	// Can't split at or beyond the clip's outer edges.
	if (t <= trimStart + EPS || t >= trimEnd - EPS) return null;
	// Can't split inside a removed range; can't split exactly on a cut edge
	// (that boundary already exists).
	for (const c of normalizeCuts(shape.cuts)) {
		if (t > c.start - EPS && t < c.end + EPS) return null;
	}
	// Already split here.
	if (splitPoints.some((p) => Math.abs(p - t) <= EPS)) return null;
	return [...splitPoints, t].sort((a, b) => a - b);
}

export interface DeletePlan {
	/** The range to remove (becomes a `manual` cut). */
	cut: { start: number; end: number };
	/** Split points that survive the delete (interior ones are pruned). */
	splitPoints: number[];
}

/**
 * Plan a ripple-delete of `seg`: the segment's range becomes a cut, and any
 * split points that sat inside (or on the edges of) the deleted range are
 * pruned, since they no longer separate two kept segments once the range is gone.
 * The caller applies the cut (the existing time-remap closes the gap) and
 * stores the returned split points.
 */
export function planDeleteSegment(
	seg: Segment,
	splitPoints: number[],
): DeletePlan {
	return {
		cut: { start: seg.start, end: seg.end },
		splitPoints: splitPoints.filter(
			(p) => p < seg.start - EPS || p > seg.end + EPS,
		),
	};
}

File: crates/recast-time/src/segments.rs
#[cfg(feature = "serde")]
use serde::{Deserialize, Serialize};

use crate::cuts::{normalize_cuts, Cut};
use crate::EPS;

#[derive(Debug, Clone, Copy, PartialEq)]
#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
pub struct Segment {
    pub start: f64,
    pub end: f64,
    pub index: usize,
}

impl Segment {
    pub fn duration(&self) -> f64 {
        self.end - self.start
    }
}

#[derive(Debug, Clone, Default)]
pub struct ClipShape {
    pub trim_start: f64,
    pub trim_end: f64,
    pub cuts: Vec<Cut>,
    pub split_points: Vec<f64>,
}

pub fn derive_segments(shape: &ClipShape) -> Vec<Segment> {
    if shape.trim_end - shape.trim_start <= EPS {
        return Vec::new();
    }

    let cuts: Vec<Cut> = normalize_cuts(&shape.cuts)
        .into_iter()
        .filter(|c| c.end > shape.trim_start && c.start < shape.trim_end)
        .collect();

    let mut kept: Vec<(f64, f64)> = Vec::new();
    let mut cursor = shape.trim_start;
    for c in cuts {
        let cut_start = c.start.max(shape.trim_start);
        let cut_end = c.end.min(shape.trim_end);
        if cut_start - cursor > EPS {
            kept.push((cursor, cut_start));
        }
        cursor = cursor.max(cut_end);
    }
    if shape.trim_end - cursor > EPS {
        kept.push((cursor, shape.trim_end));
    }

    let mut segments: Vec<Segment> = Vec::new();
    for (start, end) in kept {
        let mut inside: Vec<f64> = shape
            .split_points
            .iter()
            .copied()
            .filter(|p| *p > start + EPS && *p < end - EPS)
            .collect();
        inside.sort_by(f64::total_cmp);
        let mut from = start;
        for p in inside {
            segments.push(Segment {
                start: from,
                end: p,
                index: segments.len(),
            });
            from = p;
        }
        segments.push(Segment {
            start: from,
            end,
            index: segments.len(),
        });
    }
    segments
}

#[derive(Debug, Clone, Copy, PartialEq)]
pub struct Seam {
    pub gap_start: f64,
    pub gap_end: f64,
    pub removed: f64,
}

pub fn derive_seams(segments: &[Segment]) -> Vec<Seam> {
    segments
        .windows(2)
        .filter_map(|pair| {
            let gap = pair[1].start - pair[0].end;
            (gap > EPS).then_some(Seam {
                gap_start: pair[0].end,
                gap_end: pair[1].start,
                removed: gap,
            })
        })
        .collect()
}

/// Right-biased at an interior boundary (NLE convention); the clip's final
/// instant belongs to the last segment.
pub fn segment_at(segments: &[Segment], t: f64) -> Option<Segment> {
    if let Some(seg) = segments
        .iter()
        .copied()
        .find(|s| t >= s.start - EPS && t < s.end - EPS)
    {
        return Some(seg);
    }
    segments
        .last()
        .copied()
        .filter(|last| (t - last.end).abs() <= EPS)
}

pub fn plan_split(t: f64, shape: &ClipShape) -> Option<Vec<f64>> {
    if t <= shape.trim_start + EPS || t >= shape.trim_end - EPS {
        return None;
    }
    if normalize_cuts(&shape.cuts)
        .iter()
        .any(|c| t > c.start - EPS && t < c.end + EPS)
    {
        return None;
    }
    if shape.split_points.iter().any(|p| (p - t).abs() <= EPS) {
        return None;
    }
    let mut out = shape.split_points.clone();
    out.push(t);
    out.sort_by(f64::total_cmp);
    Some(out)
}

#[derive(Debug, Clone, PartialEq)]
pub struct DeletePlan {
    pub cut: Cut,
    pub split_points: Vec<f64>,
}

pub fn plan_delete_segment(seg: &Segment, split_points: &[f64]) -> DeletePlan {
    DeletePlan {
        cut: Cut::new(seg.start, seg.end),
        split_points: split_points
            .iter()
            .copied()
            .filter(|p| *p < seg.start - EPS || *p > seg.end + EPS)
            .collect(),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn shape(trim_start: f64, trim_end: f64, cuts: &[(f64, f64)], splits: &[f64]) -> ClipShape {
        ClipShape {
            trim_start,
            trim_end,
            cuts: cuts.iter().map(|(s, e)| Cut::new(*s, *e)).collect(),
            split_points: splits.to_vec(),
        }
    }

    #[test]
    fn an_uncut_clip_is_one_segment() {
        let segs = derive_segments(&shape(2.0, 8.0, &[], &[]));
        assert_eq!(segs.len(), 1);
        assert_eq!((segs[0].start, segs[0].end), (2.0, 8.0));
    }

    #[test]
    fn a_cut_splits_the_clip_and_indices_stay_contiguous() {
        let segs = derive_segments(&shape(0.0, 10.0, &[(4.0, 6.0)], &[]));
        assert_eq!(segs.len(), 2);
        assert_eq!(segs[1].index, 1);
        assert_eq!((segs[0].end, segs[1].start), (4.0, 6.0));
    }

    #[test]
    fn a_split_subdivides_without_removing_time() {
        let segs = derive_segments(&shape(0.0, 10.0, &[], &[4.0]));
        assert_eq!(segs.len(), 2);
        assert_eq!(segs[0].end, segs[1].start);
    }

    #[test]
    fn split_points_inside_a_cut_or_outside_the_clip_are_ignored() {
        let segs = derive_segments(&shape(0.0, 10.0, &[(4.0, 6.0)], &[5.0, 20.0, -3.0]));
        assert_eq!(segs.len(), 2);
    }

    #[test]
    fn a_cut_straddling_the_in_point_is_clamped() {
        let segs = derive_segments(&shape(10.0, 20.0, &[(8.0, 12.0)], &[]));
        assert_eq!(segs.len(), 1);
        assert_eq!(segs[0].start, 12.0);
    }

    #[test]
    fn a_degenerate_trim_yields_nothing() {
        assert!(derive_segments(&shape(5.0, 5.0, &[], &[])).is_empty());
    }

    #[test]
    fn seams_appear_only_where_time_was_removed() {
        let segs = derive_segments(&shape(0.0, 10.0, &[(4.0, 6.0)], &[8.0]));
        let seams = derive_seams(&segs);
        assert_eq!(seams.len(), 1);
        assert_eq!(seams[0].removed, 2.0);
    }

    #[test]
    fn segment_at_is_right_biased_on_a_boundary() {
        let segs = derive_segments(&shape(0.0, 10.0, &[], &[4.0]));
        assert_eq!(segment_at(&segs, 4.0).map(|s| s.index), Some(1));
        assert_eq!(segment_at(&segs, 10.0).map(|s| s.index), Some(1));
        assert!(segment_at(&segs, 11.0).is_none());
    }

    #[test]
    fn split_is_refused_at_the_edges_and_inside_a_cut() {
        let s = shape(0.0, 10.0, &[(4.0, 6.0)], &[]);
        assert!(plan_split(0.0, &s).is_none());
        assert!(plan_split(10.0, &s).is_none());
        assert!(plan_split(5.0, &s).is_none());
        assert!(plan_split(4.0, &s).is_none());
    }

    #[test]
    fn split_is_refused_where_one_already_exists() {
        let s = shape(0.0, 10.0, &[], &[4.0]);
        assert!(plan_split(4.0, &s).is_none());
        assert_eq!(plan_split(2.0, &s), Some(vec![2.0, 4.0]));
    }

    #[test]
    fn deleting_a_segment_prunes_the_splits_inside_it() {
        let seg = Segment {
            start: 4.0,
            end: 8.0,
            index: 1,
        };
        let plan = plan_delete_segment(&seg, &[2.0, 4.0, 6.0, 8.0, 9.0]);
        assert_eq!(plan.cut, Cut::new(4.0, 8.0));
        assert_eq!(plan.split_points, vec![2.0, 9.0]);
    }
}
